[BLOG] add 'Scaling NumPy on Free-Threaded Python` post (#972)
Quansight/Quansight-website: apps/labs/posts/scaling-numpy-on-free-threaded-python.md

question](https://stackoverflow.com/questions/79851420/multithreading-becomes-much-slower-than-multiprocessing-in-free-threaded-python)
where a user reported that a NumPy workload using `ThreadPoolExecutor` was
significantly _slower_ than the equivalent `ProcessPoolExecutor` version on the
free-threaded build of CPython. This was tracked as
[numpy/numpy#30494](https://github.com/numpy/numpy/issues/30494).

The reproducer is small and representative of ufunc heavy workloads: each
worker takes an array, applies a handful of `np.sin` and `np.cos` calls in a
loop, and reduces the result. There is no shared mutable state between
workers, so in principle this should scale linearly with the number of cores.
In practice, multi-threading was up to 2x _slower_ than multi-processing
on the same machine.

The free-threaded build removes the GIL, but removing the GIL is not enough on
its own. Profiling the reproducer revealed several hidden bottlenecks in NumPy
and CPython that were masked by the GIL on the regular build but became severe
scaling issues on the free-threaded build. These scaling bottlenecks fall
primarily into three categories:

1. Lock contention.

> 2. Refcount contention on shared global objects.

Reviewer: ```suggestion
2. Reference count contention on shared global objects.
```

A little less jargony, although you explain it later.

3. Contention in the memory allocator.

<figure style={{ textAlign: 'center' }}>
  <img
    src="/posts/scaling-numpy-on-free-threaded-python/before.png"
    alt="Flame graph of the reproducer before any fixes."
    style={{position:'relative'}}
  />
  <figcaption>Flame graph of the reproducer before any fixes.</figcaption>
</figure>

> 

Reviewer: Consider adding a section here explaining how to read flame graphs (or linking somewhere that does) and explaining how you actually debug stuff like this. Something like "first, run the script under samply, look at a compute thread, and try to find things in the flame graph that are taking a substantial fraction of runtime compared with calling a low-level `sin` or `cos` implementation, then try to fix the issue, and run samply again".

Maybe also talk about how you can compare with the multiprocessing version in the same samply output and this also helps spot things that are causing overhead.

Also worth emphasizing that after you fix a scaling bottleneck, that will only improve scaling until the next-worse bottleneck is fixed, so this is a very iterative process and it's not necessarily unusual that there were several different issues that all needed to be fixed to get close-to-ideal scaling.

Also maybe worth talking a little about Amdahl's Law and how any time spent running on only one core (e.g. with a global lock held) necessarily means that there is an ultimate limit to multithreaded scaling.

Reviewer: On second thought this content can live elsewhere.

## The bottlenecks and the fixes

### 1. Lock contention in `tracemalloc`

> The first bottleneck that surfaced was in `tracemalloc` of CPython.

Reviewer: ```suggestion
The first bottleneck that surfaced was in CPython's `tracemalloc` module.
```

`tracemalloc` is a memory tracking tool which is disabled by default,

> however even though it was not enabled, it still acquired a global lock

Reviewer: ```suggestion
`tracemalloc` is a memory tracking tool which is disabled by default.
However even though it was not enabled, it still acquired a global lock
```

> on every allocation and deallocation.

Reviewer: ```suggestion
on every allocation and deallocation to check whether `tracemalloc` had been
enabled at runtime.
```

> I fixed this in CPython by avoiding locking when `tracemalloc` is disabled

Reviewer: ```suggestion
I fixed this in CPython by avoiding locking when `tracemalloc` is disabled.
```

by using atomic operations to check whether it is enabled before acquiring
the lock. This was implemented in

> [python/cpython#143065](https://github.com/python/cpython/pull/143065).

Reviewer: ```suggestion
Now, CPython uses atomic operations to check whether `tracemalloc` is
enabled, providing a lock-free fast path. This was implemented in
[python/cpython#143065](https://github.com/python/cpython/pull/143065).
```

> 

Reviewer: Maybe worth adding a paragraph here explaining how this sort of bug is easy to write and that it's often not obvious at first whether or not a lock will cause a scaling issue. Also that adding a fast path is usually how you fix scaling issues like this.

### 2. Lock contention in the ufunc dispatch cache

Every call to a ufunc such as `np.sin` or `np.cos` goes through a dispatch
cache that maps a tuple of input types to a concrete loop implementation. The
cache was previously protected by a `std::shared_mutex` (a reader-writer lock).

> Even though, it was a reader-writer lock, it still didn't scale well.

Reviewer: I would break this into two paragraphs. The first could be something like this:

> NumPy implements mathematical operations like `np.sin` as ufuncs (short for universal functions). You can read more about ufuncs in the NumPy docs [add link]. The key thing to know for this discussion is that that  ufuncs are defined in terms of the input types. NumPy implements different loops for different types. For example, the implementation for `np.add` is different when the operands for the addition operation are integers or floats. For the former, NumPy needs to use a loop based on *integer* addition while for the latter we need to use floating point multiplication. In other words, NumPy has a dispatch system where it needs to determine the correct loop given a tuple of input data types.

And then the second could explain that there's a cache that speeds this up and that the cache is global state in the NumPy implementation.

I fixed this by implementing a lock-free hashtable for the dispatch cache,
which allows for lock-free reads. The fast-path for cache hit is now just
a few atomic read operations which scales well across threads. This was
implemented in
[numpy/numpy#30593](https://github.com/numpy/numpy/pull/30593), you can
read more about the lock-free design in
[npy_hashtable.c](https://github.com/numpy/numpy/blob/main/numpy/_core/src/common/npy_hashtable.c).

> TODO: describe lock-free implementation in more detail and add a diagram.

Reviewer: Awesome, looking forward to reading this.

### 3. Reference count contention on global `PyCapsule` objects

> NumPy stores the default memory handler in a global `PyCapsule` object.

Reviewer: Similar to above, add a paragraph above this explaining why NumPy has a memory handler than lives in a PyCapsule: to allow customization.

On the GIL build, an extra `Py_INCREF`/`Py_DECREF` on these capsule objects
is essentially free. On the free-threaded build, every increment and
decrement of such objects requires atomic operations on the reference count,
and with many threads all contending on the same global capsules, this
became a severe scaling bottleneck.

> 

Reviewer: Maybe add a paragraph here or link elsewhere to explain why incref and decref is *so* much slower for shared objects on the free-threaded build and how global state is particularly susceptible to triggering slowdowns due to reference count churn.

I fixed this by making the global default memory handler immortal,
meaning that it is never freed and no reference counting operations are
performed on them at all. There was no public C API to mark an object as
immortal, so I added `PyUnstable_SetImmortal` in CPython. This was
implemented in
[python/cpython#144543](https://github.com/python/cpython/pull/144543)

> and [numpy/numpy#30826](https://github.com/numpy/numpy/pull/30826)

Reviewer: Consider breaking this into two paragraphs where the first describes what making an object immortal means and that CPython uses this a lot internally to allow things to scale better. End by explaining how before Python 3.15, there was no public API for libraries like NumPy to make things immortal. End the second paragraph by explaining how this use-case justified adding the API and that we can make this immortal on Python 3.14 by using the compat headers, which implement the operation using private C APIs with the knowledge that Python 3.15 made the functionality public.

Also consider adding a paragraph about free-threading is still new and there still might be things that are needed in the C API to allow extensions to be performant. This example shows how it's not a lost cause to get the performance win, even on released Python versions, so long as you go through the trouble to expose the functionality you want in the CPython development branch.

### 4. Module attribute lookup contention

NumPy uses module level `__getattr__` to resolve ufuncs such as `np.sin`

> and `np.cos` to their actual implementations. In CPython, the module

Reviewer: Consider adding a sentence spelling out a little more concretely that you mean literally the attribute access on the `numpy` module.

attribute lookup bytecode specialization was not enabled for modules
which defined `__getattr__` which causes the attribute lookup to follow
the slow-path of acquiring the import lock and performing the lookup.

I fixed this upstream in CPython
([python/cpython#143470](https://github.com/python/cpython/pull/143470))
by enabling bytecode specialization for module attribute lookups even

> when `__getattr__` is defined.

Reviewer: Maybe say that this fix is only available in Python 3.15 (right?) but if you hit this in 3.14, you can avoid the slowdown by avoiding doing `getattr` repeatedly in a hot loop.

### 5. Memory allocator contention

After all of the above, the only remaining contention in the flame graph was
inside `malloc` itself. The free-threaded workload was allocating memory
concurrently across many threads, and it didn't scale well because of
locking inside the system allocator implementation particularly on macOS.

The fix was twofold:

- In CPython, I changed the raw allocator APIs to use mimalloc as the
  underlying memory allocator on the free-threaded build. This was implemented

>   in [python/cpython#144916](https://github.com/python/cpython/pull/144916).

Reviewer: Add a sentence explaining how `mimalloc` is designed to have good multithreaded scaling and this completely bypasses poor multithreaded in scaling on system allocators like the one on MacOS.

> - In NumPy, I changed the array allocation APIs to use the raw allocator

Reviewer: I would maybe add a whole paragraph explaining how NumPy called `malloc` and `free` directly but that CPython offers low-level "raw" allocator routines NumPy could use instead. Doing the change to using the raw allocator allows NumPy to use `mimalloc` instead of the system allocator without knowing anything about `mimalloc`. Also add a link to the CPython memory allocation docs.

>   instead system allocator. This was implemented in

Reviewer: ```suggestion
  instead of the system allocator. This was implemented in
```

  [numpy/numpy#30846](https://github.com/numpy/numpy/pull/30846).

## Benchmarks

Here are the benchmark results comparing the performance of the
multi-threaded reproducer on the free-threaded build before and after all of
the above fixes on a 32 core linux machine:

<figure style={{ textAlign: 'center' }}>
  <img
    src="/posts/scaling-numpy-on-free-threaded-python/so_benchmark.png"
    alt="Benchmark result."
    style={{position:'relative'}}
  />
  <figcaption>
    The red line represents the performance before the fixes, the green line
    represents the performance after the fixes and the black line represents
    the performance of the multi-process version.
  </figcaption>
</figure>

> Before the fixes, the multi-threaded scaled well upto 18 threads, but after

Reviewer: ```suggestion
Before the fixes, the multi-threaded case scaled well up to 18 threads, but after
```

that because of the bottlenecks described above, the performance degraded
significantly and became much slower than the multi-process version. After
the fixes, the multi-threaded version scales well across all 32 cores and is
significantly faster than the multi-process version.

## Summary

> NumPy now scales well on the free-threaded build of CPython after several

Reviewer: ```suggestion
NumPy ufuncs now scale well on the free-threaded build of CPython after several
```

read the updated entry without data races. If the cache needs to be resized,
the writer creates a new buckets table, rehashes all entries into it, and
atomically updates the pointer to the new table. The old table is kept alive
until deallocation, so readers that are still using it can finish safely
without worrying about it being freed while they are reading from it. This
design allows the dispatch cache to scale well with many threads while being
thread-safe. This was implemented in
[numpy/numpy#30593](https://github.com/numpy/numpy/pull/30593).

<figure style={{ textAlign: 'center' }}>
  <img
    src="/posts/scaling-numpy-on-free-threaded-python/hashtable.png"
    alt="Diagram of the lock-free dispatch cache. "
    style={{position:'relative'}}
  />
  <figcaption>
    The lock-free dispatch cache. Readers follow the atomic pointer to the
    current buckets table and look up entries without taking a lock. After the
    table is resized, the old table is kept alive (linked through a prev
    chain) until deallocation, so readers still using it — like Reader C — can
    finish safely. Only writers acquire the mutex.
  </figcaption>

> </figure>

Reviewer: I think the explanation above makes more sense if you move this diagram up one or two paragraphs.

Author: done

many threads are referencing the same global `PyCapsule` object for the
default memory handler.

I fixed this by making the global default memory handler immortal. An immortal
object is one that is never freed and has no reference counting performed on it
at all, which sidesteps the contention entirely. CPython already uses immortal
objects extensively for internal singletons like `None`, `True`, and small
integers, precisely because it lets them be shared across threads without
reference count contention.

The problem was that, before Python 3.15, there was no public C API for a
library like NumPy to make its own objects immortal. This use case is exactly
what justified adding one, so I added `PyUnstable_SetImmortal` in CPython. This
was implemented in
[python/cpython#144543](https://github.com/python/cpython/pull/144543)
and [numpy/numpy#30826](https://github.com/numpy/numpy/pull/30826). The API is
public from Python 3.15 onwards, but NumPy can already use it on Python 3.14
through the [pythoncapi-compat](https://github.com/python/pythoncapi-compat)
headers, which implement it using the private C APIs that back the public

> function in 3.15.

Reviewer: 3.14, right?

Author: yes

### 4. Module attribute lookup contention

> When you write `np.sin`, that is literally an attribute access on the `numpy`

Reviewer: I'd drop "literally"

acquiring the import lock and performing the lookup.

I fixed this upstream in CPython
([python/cpython#143470](https://github.com/python/cpython/pull/143470))
by enabling bytecode specialization for module attribute lookups even
when `__getattr__` is defined. This fix is available from Python 3.15 onwards.

### 5. Memory allocator contention

NumPy allocates the memory for its arrays directly from the system allocator
through `malloc` and `free`. With many threads allocating and freeing memory
concurrently, this scales poorly because some system allocators serialize
allocations behind internal locks. This is especially bad on macOS, where the
system allocator does not scale well when multiple threads are allocating and
deallocating memory.

CPython exposes its own low-level ["raw" memory allocator
routines](https://docs.python.org/3/c-api/memory.html) such as
`PyMem_RawMalloc` and `PyMem_RawFree`. However, those routines were also
calling the system allocator. The free-threaded build of CPython already uses
[mimalloc](https://github.com/microsoft/mimalloc), which is highly optimized
for multi-threaded workloads, to allocate Python objects, but NumPy was not

> using it because it was calling the system allocator directly.

Reviewer: Break this run-on up into more sentences.

The fix was twofold:

- In CPython, I changed the raw allocator APIs to use mimalloc as the
  underlying memory allocator on the free-threaded build. This was implemented
  in [python/cpython#144916](https://github.com/python/cpython/pull/144916).

- In NumPy, I changed the array allocation APIs to use CPython's raw allocator
  instead of calling the system allocator directly. Routing array allocations
  through these routines lets NumPy automatically benefit from mimalloc —
  without needing to know anything about mimalloc itself. This was implemented
  in [numpy/numpy#30846](https://github.com/numpy/numpy/pull/30846).

> ## Benchmarks

Reviewer: This is the same benchmark from stackoverflow or something else? Consider adding a sentence or two explaining the workflow you're benchmarking.

Author: yes, added
